File: src/main/scala/io/izzel/taboolib/module/db/local/SecuredFile.java
package io.izzel.taboolib.module.db.local;

import io.izzel.taboolib.util.Files;
import org.bukkit.Bukkit;
import org.bukkit.configuration.ConfigurationSection;
import org.bukkit.configuration.InvalidConfigurationException;
import org.bukkit.configuration.file.YamlConfiguration;

import java.io.File;
import java.text.SimpleDateFormat;
import java.util.Map;
import java.util.logging.Level;

/**
 * 线程安全的 YamlConfiguration 封装
 *
 * @author sky
 * @since 2020-02-28 11:14
 */
public class SecuredFile extends YamlConfiguration {

    private final Object lock = new Object();

    @Override
    public void set(String path, Object value) {
        synchronized (lock) {
            super.set(path, value);
        }
    }

    @Override
    public String saveToString() {
        synchronized (lock) {
            return super.saveToString();
        }
    }

    /**
     * 如果文件读取失败则创建备份
     * 以防出现不可逆的损伤
     */
    @Override
    public void load(File file) throws InvalidConfigurationException {
        String content = Files.readFromFile(file);
        try {
            loadFromString(content);
        } catch (InvalidConfigurationException t) {
            if (!file.getName().endsWith(".bak")) {
                Files.copy(file, new File(file.getParent(), file.getName() + "_" + new SimpleDateFormat("yyyyMMddHHmmss").format(System.currentTimeMillis()) + ".bak"));
            }
            throw t;
        }
    }

    /**
     * 如果文本读取失败则打印到日志
     * 以防出现不可逆的损伤
     */
    @Override
    public void loadFromString(String contents) throws InvalidConfigurationException {
        try {
            super.loadFromString(contents);
        } catch (InvalidConfigurationException t) {
            System.out.println("Source: \n" + contents);
            throw t;
        }
    }

    public static String dump(Object data) {
        if (data == null) {
            return "";
        }
        boolean single = false;
        YamlConfiguration dump = new YamlConfiguration();
        if (data instanceof ConfigurationSection) {
            ((ConfigurationSection) data).getValues(false).forEach(dump::set);
        } else if (data instanceof Map) {
            ((Map<?, ?>) data).forEach((k, v) -> dump.set(k.toString(), v));
        } else {
            single = true;
            dump.set("value", data);
        }
        String[] save;
        if (single) {
            save = dump.saveToString().substring("value:".length()).trim().split("\n");
        } else {
            save = dump.saveToString().split("\n");
        }
        return String.join("\n", save);
    }

    public static SecuredFile loadConfiguration(String contents) {
        SecuredFile config = new SecuredFile();
        try {
            config.loadFromString(contents);
        } catch (Throwable t) {
            Bukkit.getLogger().log(Level.SEVERE, "Cannot load source", t);
        }
        return config;
    }

    public static SecuredFile loadConfiguration(File file) {
        SecuredFile config = new SecuredFile();
        try {
            config.load(file);
        } catch (Throwable t) {
            Bukkit.getLogger().log(Level.SEVERE, "Cannot load " + file, t);
        }
        return config;
    }
}

File: module/module-configuration/src/main/kotlin/taboolib/module/configuration/SecuredFile.kt
package taboolib.module.configuration

import taboolib.common.platform.function.warning
import taboolib.library.configuration.ConfigurationSection
import taboolib.library.configuration.InvalidConfigurationException
import taboolib.library.configuration.YamlConfiguration
import java.io.File
import java.io.InputStream
import java.io.Reader
import java.nio.charset.StandardCharsets
import java.text.SimpleDateFormat

class SecuredFile : YamlConfiguration(), Configuration {

    private val lock = Any()

    override var file: File? = null

    override fun set(path: String, value: Any?) {
        synchronized(lock) { super.set(path, value) }
    }

    override fun saveToFile(file: File?) {
        synchronized(lock) { super.save(file ?: return) }
    }

    override fun saveToString(): String {
        synchronized(lock) { return super.saveToString() }
    }

    override fun loadFromFile(file: File) {
        load(file)
    }

    override fun loadFromReader(reader: Reader) {
        loadFromString(reader.readText())
    }

    override fun loadFromInputStream(inputStream: InputStream) {
        loadFromString(inputStream.readBytes().decodeToString())
    }

    override fun reload() {
        load(file ?: return)
    }

    /**
     * 如果文件读取失败则创建备份
     * 以防出现不可逆的损伤
     */
    @Throws(InvalidConfigurationException::class)
    override fun load(file: File) {
        this.file = file
        val content = file.readText(StandardCharsets.UTF_8)
        try {
            loadFromString(content)
        } catch (ex: InvalidConfigurationException) {
            if (!file.name.endsWith(".bak")) {
                file.copyTo(File(file.parent, file.name + "_" + SimpleDateFormat("yyyyMMddHHmmss").format(System.currentTimeMillis()) + ".bak"))
            }
            throw ex
        }
    }

    /**
     * 如果文本读取失败则打印到日志
     * 以防出现不可逆的损伤
     */
    @Throws(InvalidConfigurationException::class)
    override fun loadFromString(contents: String) {
        try {
            super.loadFromString(contents)
        } catch (t: InvalidConfigurationException) {
            warning("Source: \n$contents")
            throw t
        }
    }

    companion object {

        fun dumpAll(key: String, value: Any?, space: Int = 2): String {
            return if (key.contains('.')) {
                "${key.substringBefore('.')}:\n${" ".repeat(space)}${dumpAll(key.substringAfter('.'), value, space + 2)}"
            } else {
                val dump = dump(value)
                when {
                    dump.startsWith("-") -> "$key:\n$dump"
                    value is List<*> && value.isEmpty() -> "$key: []"
                    value is Map<*, *> -> if (value.isEmpty()) "$key: {}" else "$key:\n$dump"
                    value is ConfigurationSection -> "$key:\n  ${dump.replace("\n", "\n  ")}"
                    else -> "$key: $dump"
                }
            }
        }

        fun dump(data: Any?): String {
            if (data == null) {
                return ""
            }
            var single = false
            val dump = YamlConfiguration()
            when (data) {
                is ConfigurationSection -> {
                    data.getValues(false).forEach { (path, value) -> dump[path] = value }
                }
                is Map<*, *> -> {
                    data.forEach { (k, v) -> dump[k.toString()] = v }
                }
                else -> {
                    single = true
                    dump["value"] = data
                }
            }
            val save = if (single) {
                dump.saveToString().substring("value:".length).trim().split('\n').toTypedArray()
            } else {
                dump.saveToString().trim().split('\n').toTypedArray()
            }
            return java.lang.String.join("\n", *save)
        }

        fun loadConfiguration(contents: String): SecuredFile {
            val config = SecuredFile()
            try {
                config.loadFromString(contents)
            } catch (t: Throwable) {
                t.printStackTrace()
            }
            return config
        }

        fun loadConfiguration(file: File): SecuredFile {
            val config = SecuredFile()
            try {
                config.load(file)
            } catch (t: Throwable) {
                t.printStackTrace()
            }
            return config
        }
    }
}